### PR TITLE
Fix HTTP3 stress

### LINKED ISSUE #67442: [H/3] stress not running
```
Trying connect to the server [https://server:5001/.](https://server:5001/)
client_1  | Stress client could not connect to host https://server:5001/, 10 attempts remaining
client_1  | Stress client could not connect to host https://server:5001/, 9 attempts remaining
client_1  | Stress client could not connect to host https://server:5001/, 8 attempts remaining
client_1  | Stress client could not connect to host https://server:5001/, 7 attempts remaining
```

Also happened to me and @rzikm locally, now seen in the pipeline as well: https://dev.azure.com/dnceng/public/_build/results?buildId=1694516&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=1451f5f3-0108-5a08-5b92-e984b2a85bbd

Might be related to docker, docker compose or docker image update.

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #67442, #68958.

This PR builds latest MsQuic inside the http stress docker image.

It also adds special case for listening on IPv6Any to accept connections from all IP addresses (i.e. including IPv4), which does not happen by default (see https://github.com/microsoft/msquic/issues/2704).

### CODE REVIEW

**Comment by rzikm** [comment]
cc: @ManickaP

File: src/libraries/System.Net.Http/tests/StressTests/HttpStress/Dockerfile [hunk 1]
```diff
@@ -1,21 +1,39 @@
...
-RUN apt-add-repository https://packages.microsoft.com/debian/11/prod
-RUN apt-get update -y
-RUN apt-get install -y libmsquic
 RUN apt-get upgrade -y
+RUN apt-get install -y cmake clang ruby-dev gem lttng-tools libssl-dev && \
+    gem install fpm
+RUN git clone --recursive https://github.com/dotnet/msquic
+RUN cd msquic/src/msquic && \
+    mkdir build && \
+    options="-DQUIC_BUILD_TOOLS=off -DQUIC_BUILD_TEST=off -DQUIC_BUILD_PERF=off" && \
```
Comment (CarnaViire) [inline src/libraries/System.Net.Http/tests/StressTests/HttpStress/Dockerfile:15]: I'm just curious, is this line needed? the options are specified below in the call once again...

Comment (rzikm) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Dockerfile:15]: I copied the code from Manicka's branch :D but agree that it looks unnecessary

Comment (ManickaP) [reply src/libraries/System.Net.Http/tests/StressTests/HttpStress/Dockerfile:15]: No they're not. This is a remnant of a salvaged code from my personal script for building msquic :smile:

**Comment by rzikm** [comment]
The stress CI leg is failing because we run out of space when building the docker image, I filed https://github.com/dotnet/arcade/issues/9349

**Review by ManickaP — COMMENTED** [review]
LGTM in general.
Some more comments that are not necessarily prereqs for this PR:
- what if you use IPv4.Any and client tries IPv6 address? How do sockets behave?
- I ran this test:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs#L31
But I changed `IPv6Loopback` -> `IPv6Any` and it got stuck somewhere in the middle of a handshake. I didn't investigate further.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs [hunk 2]
```diff
@@ -179,35 +179,41 @@ private unsafe IPEndPoint Start(QuicListenerOptions options)
...
             }
 
+            // return the actual bound address, including a port. Since the address family may be Unspecified,
+            // we cannot use GetIPEndPointParam. We have to manually read the port from the raw QuicAddr structure.
+            // The actual address is unchanged.
             Debug.Assert(!Monitor.IsEntered(_state), "!Monitor.IsEntered(_state)");
-            return MsQuicParameterHelpers.GetIPEndPointParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_LISTENER_LOCAL_ADDRESS);
+            QuicAddr listenAddr = MsQuicParameterHelpers.GetQuicAddrParam(MsQuicApi.Api, _state.Handle, QUIC_PARAM_LISTENER_LOCAL_ADDRESS);
+            int port = BinaryPrimitives.ReadUInt16BigEndian(MemoryMarshal.AsBytes(MemoryMarshal.CreateSpan(ref listenAddr.Ipv4.sin_port, 1)));
+            return new IPEndPoint(listenEndPoint.Address, port);
```
[the remark this answers is not in the document]
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:216]: I have checked MsQuic source code, and they only change the port (if it was 0 or AF was Unspecified)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs [hunk 3]
```diff
@@ -179,35 +179,41 @@ private unsafe IPEndPoint Start(QuicListenerOptions options)
...
+                    &address), "ListenerStart failed");
             }
             catch
             {
                 _stateHandle.Free();
                 throw;
             }
 
+            // return the actual bound address, including a port. Since the address family may be Unspecified,
+            // we cannot use GetIPEndPointParam. We have to manually read the port from the raw QuicAddr structure.
```
[the remark this answers is not in the document]
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicListener.cs:211]: you mean add optional parameter to GetIPEndPointParam? that seems to work. Thanks.

**Comment by rzikm** [comment]
> I ran this test:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs#L31
But I changed IPv6Loopback -> IPv6Any and it got stuck somewhere in the middle of a handshake. I didn't investigate further.

The test tries to connect to `listener.ListenEndPoint`, so if you didn't change that, then you tried to connect to IPv6Any, which dos not work.

> what if you use IPv4.Any and client tries IPv6 address? How do sockets behave?

Even if you use `DualMode` sockets, then you can't connect via IPv6 if server listens on IPv4Any

**Comment by ManickaP** [comment]
> > I ran this test:
> > https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Quic/tests/FunctionalTests/QuicListenerTests.cs#L31
> > But I changed IPv6Loopback -> IPv6Any and it got stuck somewhere in the middle of a handshake. I didn't investigate further.
> 
> The test tries to connect to `listener.ListenEndPoint`, so if you didn't change that, then you tried to connect to IPv6Any, which dos not work.

Works with sockets though, I just tested it. It'll connect to IPv6 loopback if you pass IPv6Any to client connect, the same for IPv4Any.

**Comment by ManickaP** [comment]
Your code runs for me as well:
```
Listening on [::]:5004
IsDualMode: True
IPV6_V6Only: 0
clientIpv6.DualMode == False
IsDualMode: True
IPV6_V6Only: 0
```

**Comment by rzikm** [comment]
That would be probably a difference between Windows and Linux networking stack. It runs for me in WSL

**Review by ManickaP — APPROVED** [review]
LGTM, we can finish the discussion about sockets offline :smile: Thanks!

**Pull request merged**